compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -2815,9 +2815,179 @@ class JSCodeGen()(using genCtx: Context) {
+          genReflectiveCall(tree, isSelectDynamic = true)
+      case REFLECT_SELECTABLE_APPLYDYN =>
+        // scala.reflect.Selectable.applyDynamic
+        if (currentClassSym.get == jsdefn.ReflectSelectableClass)
+          genNormalApply(tree, isStat)
+        else
+          genReflectiveCall(tree, isSelectDynamic = false)
     }
   }
 
+  /** Gen the SJSIR for a reflective call.
+   *
+   *  Reflective calls are calls to a structural type field or method that
+   *  involve a reflective Selectable. They look like the following in source
+   *  code:
+   *  {{{
+   *  import scala.reflect.Selectable.reflectiveSelectable
+   *
+   *  type Structural = {
+   *    val foo: Int
+   *    def bar(x: Int, y: String): String
+   *  }
+   *
+   *  val structural: Structural = new {
+   *    val foo: Int = 5
+   *    def bar(x: Int, y: String): String = x.toString + y
+   *  }
+   *
+   *  structural.foo
+   *  structural.bar(6, "hello")
+   *  }}}
+   *
+   *  After expansion by the Scala 3 rules for structural member selections and
+   *  calls, they look like
+   *
+   *  {{{
+   *  reflectiveSelectable(structural).selectDynamic("foo")
+   *  reflectiveSelectable(structural).applyDynamic("bar",
+   *    ClassTag(classOf[Int]), ClassTag(classOf[String])
+   *  )(
+   *    6, "hello"
+   *  )
+   *  }}}
+   *
+   *  and eventually reach the back-end as
+   *
+   *  {{{
+   *  reflectiveSelectable(structural).selectDynamic("foo") // same as above
+   *  reflectiveSelectable(structural).applyDynamic("bar",
+   *    wrapRefArray([ ClassTag(classOf[Int]), ClassTag(classOf[String]) : ClassTag ]
+   *  )(
+   *    genericWrapArray([ Int.box(6), "hello" : Object ])
+   *  )
+   *  }}}
+   *
+   *  If we use the deprecated `import scala.language.reflectiveCalls`, the
+   *  wrapper for the receiver `structural` are the following instead:
+   *
+   *  {{{
+   *  reflectiveSelectableFromLangReflectiveCalls(structural)(
+   *    using scala.languageFeature.reflectiveCalls)
+   *  }}}
+   *
+   *  (in which case we don't care about the contextual argument).
+   *
+   *  In SJSIR, they must be encoded as follows:
+   *
+   *  {{{
+   *  structural.foo;R()
+   *  structural.bar;I;Ljava.lang.String;R(
+   *    Int.box(6).asInstanceOf[int],
+   *    "hello".asInstanceOf[java.lang.String]
+   *  )
+   *  }}}
+   *
+   *  This means that we must deconstruct the elaborated calls to recover:
+   *
+   *  - the original receiver `structural`
+   *  - the method name as a compile-time string `foo` or `bar`
+   *  - the `tp: Type`s that have been wrapped in `ClassTag(classOf[tp])`, as a
+   *    compile-time List[Type], from which we'll derive `jstpe.Type`s for the
+   *    `asInstanceOf`s and `jstpe.TypeRef`s for the `MethodName.reflectiveProxy`
+   *  - the actual arguments as a compile-time `List[Tree]`
+   *
+   *  Virtually all of the code in `genReflectiveCall` deals with recovering
+   *  those elements. Constructing the IR Tree is the easy part after that.
```

Comment:
Using a thing like `scala.reflect.Selectable` seems like a neat trick to easily implement reflective calls on backends that support reflection.

However, here it seems that this trick is now doing the entire compiler a disservice: We have ~100 LOC simply reconstructing things and then a single LOC actually issuing the call. (and additionally failures that IIUC simply cannot happen with reflective calls).

Have you considered introducing a form of transient tree and doing the translation to a call to `Selectable` later? (I realize that the call relies on other transformations that happen earlier, so it might be worse overall).

Reply:
I agree that the situation is very inconvenient. But the desugaring that is now done in Typer is actually *specced* as is now. It's not "a trick". It is the normal way of compiling calls to members of structural types. `reflectiveSelectable` is just *one* possible implicit that can provide an implementation for those, but there can be other, user-defined implementation (that don't really on Java reflection, but on an internal Map, for example).

The reference is here: https://dotty.epfl.ch/docs/reference/changed-features/structural-types.html

We need to support `reflectiveSelectable` because it corresponds to Scala 2 reflective calls, and it's the only way that our back-end would emit calls to reflective proxies (which we really need in some cases; we use them in the JDK implem in Scala.js itself for example).

Users can implement other `Selectable`s not relying on Java run-time reflection, and those would also work.

So we *can't* introduce a form of transient tree to be desugared later. The desugaring is by spec and requires interactions with other tasks of typechecking, including implicit resolution. It's impossible to delay that.

Reply:
Hmmm... I see. So is the expectation that any custom `Selectable` that needs compile time reflection needs to do this ceremony? The page you linked doesn't go into details about that.

Further, if that is the case, shouldn't we introduce *another* `Selectable` for Scala.js? The main issue I see here is that with the current specification, it would be allowed to use `reflectiveSelectable` on its own, without structural types. However, such use would immediately lead to non-portable code. Maybe we should make this more obvious?

Reply:
I realize this is probably a larger discussion, so you might want to go ahead with this PR. I still think it is worth thinking about this a bit more.